File: packages/ts/src/x25519.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { x25519 } from "@noble/curves/ed25519.js";
import { ReallyMeCryptoError } from "./errors.js";

/**
 * X25519 key agreement backed by @noble/curves — the same pinned
 * implementation the TypeScript conformance lane proves vectors against.
 *
 * The package returns the raw 32-byte Diffie-Hellman output. Higher-level
 * protocols must bind it through their own KDF transcript; this primitive does
 * not apply HKDF implicitly because HPKE, MLS, and ratchets label transcripts
 * differently.
 */
export const X25519_SECRET_KEY_LENGTH = 32;
export const X25519_PUBLIC_KEY_LENGTH = 32;
export const X25519_SHARED_SECRET_LENGTH = 32;

export const ReallyMeX25519 = {
  /** Generates a random X25519 keypair: 32-byte public key, 32-byte secret. */
  generateKeyPair(): { publicKey: Uint8Array; secretKey: Uint8Array } {
    const secretKey = x25519.utils.randomSecretKey();
    return {
      publicKey: x25519.getPublicKey(secretKey),
      secretKey,
    };
  },

  /** Derives the X25519 keypair from a 32-byte secret. */
  deriveKeyPair(secretKey: Uint8Array): { publicKey: Uint8Array; secretKey: Uint8Array } {
    return {
      publicKey: this.derivePublicKey(secretKey),
      secretKey: secretKey.slice(),
    };
  },

  /** Derives the 32-byte X25519 public key from a 32-byte secret. */
  derivePublicKey(secretKey: Uint8Array): Uint8Array {
    if (secretKey.length !== X25519_SECRET_KEY_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      return x25519.getPublicKey(secretKey);
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },

  /** Derives the raw 32-byte X25519 shared secret. */
  deriveSharedSecret(publicKey: Uint8Array, secretKey: Uint8Array): Uint8Array {
    if (
      publicKey.length !== X25519_PUBLIC_KEY_LENGTH ||
      secretKey.length !== X25519_SECRET_KEY_LENGTH
    ) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      const sharedSecret = x25519.getSharedSecret(secretKey, publicKey);
      if (!sharedSecret.some((byte) => byte !== 0)) {
        throw new ReallyMeCryptoError("invalid-input");
      }
      return sharedSecret;
    } catch (error) {
      if (error instanceof ReallyMeCryptoError) {
        throw error;
      }
      throw new ReallyMeCryptoError("invalid-input");
    }
  },
} as const;

File: crates/crypto/src/x25519.rs
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

//! X25519 root facade with key agreement routed through the operation layer.

use crypto_core::{Algorithm, CryptoError};
use zeroize::Zeroizing;

use crate::key_agreement_error::{
    crypto_error_from_derive_shared_secret_operation_error,
    crypto_error_from_key_generation_operation_error,
};

pub use crypto_x25519::{decode_public_key, encode_public_key};

/// Generate an X25519 keypair through the key-agreement operation owner.
pub fn generate_x25519_keypair() -> Result<(Vec<u8>, Zeroizing<Vec<u8>>), CryptoError> {
    crate::operations::key_agreement::generate_key_pair(Algorithm::X25519)
        .map(|key_pair| (key_pair.public_key, key_pair.secret_key))
        .map_err(crypto_error_from_key_generation_operation_error)
}

/// Reconstruct an X25519 keypair from an existing 32-byte secret seed.
pub fn generate_x25519_keypair_from_seed(
    seed: &[u8; 32],
) -> Result<(Vec<u8>, Zeroizing<Vec<u8>>), CryptoError> {
    crate::operations::key_agreement::derive_key_pair(Algorithm::X25519, seed)
        .map(|key_pair| (key_pair.public_key, key_pair.secret_key))
        .map_err(crypto_error_from_key_generation_operation_error)
}

/// Derive the raw X25519 shared secret through the key-agreement operation owner.
pub fn derive_x25519_shared_secret(
    secret_key: &[u8],
    public_key: &[u8],
) -> Result<Zeroizing<Vec<u8>>, CryptoError> {
    crate::operations::key_agreement::derive_shared_secret(
        Algorithm::X25519,
        secret_key,
        public_key,
    )
    .map_err(crypto_error_from_derive_shared_secret_operation_error)
}
